### PR TITLE
[refactor] Migrate Schedules Table to typescript

### PR DESCRIPTION
**Pull request by muhsinkamil**
Relates to  https://github.com/actualbudget/actual/issues/1483

- [x] Migrate Schedules Table and its child components to ts

### CODE REVIEW

File: packages/desktop-client/src/components/schedules/SchedulesTable.tsx [hunk 1]
```diff
@@ -274,7 +311,7 @@ export function SchedulesTable({
         </Row>
       );
     }
-    return renderSchedule({ item });
+    return renderSchedule({ schedule: item as ScheduleEntity });
```
Comment (muhsinkamil) [inline packages/desktop-client/src/components/schedules/SchedulesTable.tsx:314]: Unfortunately, had to cast it as the items as a mix of `ScheduleEntity` & `CompletedScheduleItem`

But not very bad tho, as the implementation still remains the same., 
( except I renamed `type === 'show-completed'    ->       id === 'show-completed `   to satisfy ts 😃 )

File: packages/desktop-client/src/components/schedules/StatusBadge.tsx [hunk 2]
```diff
@@ -94,7 +87,6 @@ export function StatusBadge({ status, style }: StatusBadgeProps) {
         flexDirection: 'row',
         alignItems: 'center',
         flexShrink: 0,
-        ...style,
```
Comment (muhsinkamil) [inline packages/desktop-client/src/components/schedules/StatusBadge.tsx:97]: Seems this is not used anywhere, hence removing it.,

File: packages/desktop-client/src/components/schedules/SchedulesTable.tsx [hunk 3]
```diff
@@ -204,42 +241,42 @@ export function SchedulesTable({
...
         <Field width={120} name="status" style={{ alignItems: 'flex-start' }}>
-          <StatusBadge status={statuses.get(item.id)} />
+          <StatusBadge status={statuses.get(schedule.id)} />
         </Field>
-        <ScheduleAmountCell amount={item._amount} op={item._amountOp} />
+        <ScheduleAmountCell amount={schedule._amount} op={schedule._amountOp} />
         {!minimal && (
           <Field width={80} style={{ textAlign: 'center' }}>
-            {item._date && item._date.frequency && (
+            {schedule._date && (schedule._date as any).frequency && (
```
Comment (muhsinkamil) [inline packages/desktop-client/src/components/schedules/SchedulesTable.tsx:270]: Had to do as any, as defining a type to `_date` would require more additional changes, I plan to look into it when I do `EditSchedule` component as I can better understand how this _date is set when making new schedule

Comment (MatissJanis) [reply packages/desktop-client/src/components/schedules/SchedulesTable.tsx:289]: Could you please add a `// eslint-disable-next-line @typescript-eslint/no-explicit-any` on the line above so the linter would pass?

Comment (muhsinkamil) [reply packages/desktop-client/src/components/schedules/SchedulesTable.tsx]: I looked into it and added a proper type 👍

File: packages/loot-core/src/client/data-hooks/schedules.tsx [hunk 4]
```diff
@@ -23,13 +26,15 @@ function loadStatuses(schedules: ScheduleEntity[], onData) {
...
   schedules: ScheduleEntity[];
-  statuses: Record<string, ReturnType<typeof getStatus>>;
+  statuses: ScheduleStatuses;
 } | null;
-export function useSchedules({ transform }: UseSchedulesArgs = {}) {
-  let [data, setData] = useState<UseSchedulesReturnType | null>(null);
+export function useSchedules({
+  transform,
+}: UseSchedulesArgs = {}): UseSchedulesReturnType {
+  const [data, setData] = useState<UseSchedulesReturnType>(null);
```
Comment (MatissJanis) [inline packages/loot-core/src/client/data-hooks/schedules.tsx:34]: **🔨 warning:** ‏missing `null` 
```suggestion
  const [data, setData] = useState<UseSchedulesReturnType | null>(null);
```

File: packages/desktop-client/src/components/schedules/SchedulesTable.tsx [hunk 5]
```diff
@@ -49,22 +85,33 @@ function OverflowMenu({ schedule, status, onAction }) {
           onClose={() => setOpen(false)}
         >
           <Menu
-            onMenuSelect={name => {
+            onMenuSelect={(name: ScheduleItemAction) => {
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/schedules/SchedulesTable.tsx:88]: **💬 suggestion:** ‏we should be able to automatically infer this type from `Menu` arguments

File: packages/loot-core/src/client/data-hooks/accounts.tsx [hunk 6]
```diff
@@ -11,7 +12,7 @@ function useAccounts() {
 let AccountsContext = createContext(null);
 
 export function AccountsProvider({ children }) {
-  let data = useAccounts();
+  const data: AccountEntity[] = useAccounts();
```
Comment (MatissJanis) [inline packages/loot-core/src/client/data-hooks/accounts.tsx:15]: **💬 suggestion:** ‏can you move this type to `useAccounts` hook instead?
```
function useAccounts(): AccountEntity[] {
```

File: packages/loot-core/src/client/data-hooks/payees.tsx [hunk 7]
```diff
@@ -11,7 +12,7 @@ function usePayees() {
 let PayeesContext = createContext(null);
 
 export function PayeesProvider({ children }) {
-  let data = usePayees();
+  const data: PayeeEntity[] = usePayees();
```
Comment (MatissJanis) [inline packages/loot-core/src/client/data-hooks/payees.tsx:15]: Same thing here

File: packages/desktop-client/src/components/schedules/index.tsx [hunk 8]
```diff
@@ -88,9 +91,6 @@ export default function Schedules() {
           onSelect={onEdit}
           onAction={onAction}
           style={{ backgroundColor: theme.tableBackground }}
-          // @todo: Remove following props after typing SchedulesTable
-          minimal={undefined}
-          tableStyle={undefined}
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/schedules/index.tsx:93]: **👏 praise:** ‏thanks!

File: packages/desktop-client/src/components/schedules/SchedulesTable.tsx [hunk 9]
```diff
@@ -21,10 +26,41 @@ import DisplayId from '../util/DisplayId';
...
+type CompletedScheduleItem = { id: 'show-completed' };
+type SchedulesTableItem = ScheduleEntity | CompletedScheduleItem;
+
+export enum ScheduleItemAction {
+  PostTransaction = 'post-transaction',
+  SkipSchedule = 'skip',
+  CompleteSchedule = 'complete',
+  RestartSchedule = 'restart',
+  DeleteSchedule = 'delete',
+}
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/schedules/SchedulesTable.tsx:50]: **💬 suggestion:** ‏we generally use union types everywhere and not enums. Could you please move this back?

**Review by MatissJanis — CHANGES_REQUESTED** [review]
Love the changes! Soon we'll be fully ported over to TS :)

File: packages/loot-core/src/client/data-hooks/schedules.tsx (hunk 4 above)
Comment (muhsinkamil) [reply packages/loot-core/src/client/data-hooks/schedules.tsx:34]: Actually, `UseSchedulesReturnType` has null as 'or' type

File: packages/desktop-client/src/components/schedules/SchedulesTable.tsx (hunk 5 above)
Comment (muhsinkamil) [reply packages/desktop-client/src/components/schedules/SchedulesTable.tsx:120]: The inference is a `string`, I added a type to enforce it to a more strict one. If you think, it is too much of a type definitions explicitly defined, I can remove it so to infer to `string`

File: packages/loot-core/src/client/data-hooks/schedules.tsx (hunk 4 above)
Comment (MatissJanis) [reply packages/loot-core/src/client/data-hooks/schedules.tsx:34]: Oh it already has `null`. Thanks for checking!

**Review by MatissJanis — APPROVED** [review]
Amazing

**Pull request merged**